celo-org/celo-monorepo
# Add FeeCurrencyDirectory to Anvil migrations

### Description

Adds the new `FeeCurrencyDirectory` contract to the Anvil migration script.

### Changes

1. deploys and proxies the `FeeCurrencyDirectory` contract
2. registers `FeeCurrencyDirectory` in the `Registry`
3. adds three existing cStables (cUSD, cEUR, and cREAL) to the `FeeCurrencyDirectory`

### Other changes

1. adds `is IOracle` to the class definition in `SortedOracles`
4. removes a duplicate import in `Migrations.s.sol`

### Tested

Yes, tested locally:
- [x] get the `FeeCurrencyDirectory` address from the `Registry`
- [x] get registered currencies from the `FeeCurrencyDirectory` contract
- [x] get the exchange rate for a currency from the `FeeCurrencyDirectory` contract

<details>
  <summary>Steps to reproduce tests locally</summary>

Requirement:
1. Have foundry installed
2. Be in the `packages/protocol` directory

Tests:

1. Run migration script (in a different Terminal)

    ```sh
    $ ./migrations_sol/create_and_migrate_anvil_devchain.sh
    ```

1. Check that Anvil is running

    ```sh
    $ nc -z localhost 8546
    Connection to localhost port 8546 [tcp/*] succeeded!
    ```
    
    Source: [start_anvil.sh](https://github.com/celo-org/celo-monorepo/blob/a36eae5bc2c9c2071fb942199b1ba265f3937c17/packages/protocol/migrations_sol/start_anvil.sh#L7)

    Anvil is running ✅

1. get the `FeeCurrencyDirectory` address from the `Registry`:

    ```sh
    export REGISTRY_ADDRESS="0x000000000000000000000000000000000000ce10"
    export ANVIL_PORT=8546
    cast call \
    $REGISTRY_ADDRESS \
    "getAddressForStringOrDie(string calldata identifier)" \
    "FeeCurrencyDirectory" \
    --rpc-url=http://127.0.0.1:$ANVIL_PORT
    
    0x00000000000000000000000042fe5a2a61ed9705eb2f08a04a58ceb606d22f6a
    ```
    
    ```sh
    cast abi-decode \
    "getAddressForStringOrDie(string calldata identifier)(address)" \
    "0x00000000000000000000000042fe5a2a61ed9705eb2f08a04a58ceb606d22f6a"
    
    0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a
    ```

    `FeeCurrencyDirectory` is registered in the `Registry` ✅

1. get registered currencies from the `FeeCurrencyDirectory` contract

    ```sh
    export FEECURRENCYDIRECTORY_PROXY_ADDRESS=0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a
    export ANVIL_PORT=8546
    
    cast call \
    $FEECURRENCYDIRECTORY_PROXY_ADDRESS \
    "getCurrencies()" \
    --rpc-url=http://127.0.0.1:$ANVIL_PORT
    
    0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003000000000000000000000000e6774be4e5f97db10cafb4c00c74cfbdcdc434d9000000000000000000000000b7a33b4ad2b1f6b0a944232f5c71798d27ad92720000000000000000000000002a3733dbc31980f02b12135c809b5da33bf3a1e9
    ```
    
    ```sh
    cast abi-decode \
    "getCurrencies()(address[] memory)" \
    "0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003000000000000000000000000e6774be4e5f97db10cafb4c00c74cfbdcdc434d9000000000000000000000000b7a33b4ad2b1f6b0a944232f5c71798d27ad92720000000000000000000000002a3733dbc31980f02b12135c809b5da33bf3a1e9"
    
    [0xe6774BE4E5f97dB10cAFB4c00C74cFbdCDc434D9, 0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272, 0x2A3733dBc31980f02b12135C809b5da33BF3a1e9]
    ```

    3 currencies are registered in the `FeeCurrencyDirectory` ✅ As expected, the registered currencies are:

    cUSD ✅
    
    ```sh
    cast call \
    0xe6774BE4E5f97dB10cAFB4c00C74cFbdCDc434D9 \
    "name()" \
    --rpc-url=http://127.0.0.1:8546
    0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000b43656c6f20446f6c6c6172000000000000000000000000000000000000000000
    ```
    
    ```sh
    cast abi-decode \
    "name()(string memory)" \
    "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000b43656c6f20446f6c6c6172000000000000000000000000000000000000000000"
    "Celo Dollar"
    ```
    
    cEUR ✅
    
    ```sh
    cast call \
    0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272 \
    "name()" \
    --rpc-url=http://127.0.0.1:8546
    0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000943656c6f204575726f0000000000000000000000000000000000000000000000
    ```
    
    ```sh
    cast abi-decode \
    "name()(string memory)" \
    "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000943656c6f204575726f0000000000000000000000000000000000000000000000"
    "Celo Euro"
    ```
    
    cREAL ✅
    
    ```sh
    cast call \
    0x2A3733dBc31980f02b12135C809b5da33BF3a1e9 \
    "name()" \
    --rpc-url=http://127.0.0.1:8546
    0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343656c6f204272617a696c69616e205265616c00000000000000000000000000
    ```
    
    ```sh
    cast abi-decode \
    "name()(string memory)" \
    "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343656c6f204272617a696c69616e205265616c00000000000000000000000000"
    "Celo Brazilian Real"
    ```

1. Get exchange rate for a currency:

    ```sh
    $ cast call \
    0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a \
    "getExchangeRate(address)(uint256, uint256)" \
    0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272 \
    --rpc-url=http://127.0.0.1:8546
    1000000000000000000000000 [1e24]
    1000000000000000000000000 [1e24]
    ```
    
    Calling the `getExchangeRate` function on a registered currency returns an exchange rate ✅

1. Terminate Anvil:

    ```sh
    $ kill $(lsof -i tcp:8546 | tail -n 1 | awk '{print $2}')
    ```
    
    Source: [start_anvil.sh](https://github.com/celo-org/celo-monorepo/blob/a36eae5bc2c9c2071fb942199b1ba265f3937c17/packages/protocol/migrations_sol/start_anvil.sh#L9)

</details>


### Related issues

- Fixes #10991

### Backwards compatibility

Yes, only adds to the migration script, doesn't alter existing expectations and behaviour.

### Documentation

No docs changes
